experiments/MoarCSets/SInstances.jl

```diff
@@ -0,0 +1,406 @@
+    :adom => ADom
+    :acodom => ACodom
+    _ => invoke(Base.getproperty, Tuple{Type,Symbol}, AD, i)
+  end
+end
+
+function ad_data_num(AD::Type{T}, data::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Data .== data)::Int
+end
+
+function ad_attr_num(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  findfirst(Attr .== attr)::Int
+end
+
+function ad_dom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  CD.ob[ADom[ad_attr_num(AD,attr)]]
+end
+
+function ad_codom(AD::Type{T}, attr::Symbol) where
+  {CD,Data,Attr,ADom,ACodom,T <: AttrDesc{CD,Data,Attr,ADom,ACodom}}
+  Data[ACodom[ad_attr_num(AD,attr)]]
+end
+
+function make_index(CD::Type{<:CatDesc},AD::Type{<:AttrDesc},Ts::Type{<:Tuple},Idxed::Tuple)
+    function make_idx(name)
+      if name ∈ CD.hom
+        Vector{Vector{Int}}()
+      elseif name ∈ AD.attr
+        Dict{Ts.parameters[ad_codom(AD,name)],Vector{Int}}()
+      else
+        error("invalid index parameter")
+      end
+    end
+  NamedTuple{Idxed}(Tuple(make_idx(name) for name in Idxed))
+end
+
+function table_desc(::Type{CatDesc{Ob,Hom,Dom,Codom}},
+                    ::Type{AttrDesc{CatDesc{Ob,Hom,Dom,Codom},Data,Attr,ADom,ACodom}},
+                    ts::Tuple) where {Ob,Hom,Dom,Codom,Data,Attr,ADom,ACodom}
+  cols = [Tuple{Symbol,DataType}[] for ob in Ob]
+  for i in 1:length(Hom)
+    push!(cols[Dom[i]], (Hom[i],Int))
+  end
+  for i in 1:length(Attr)
+    push!(cols[ADom[i]], (Attr[i],ts[ACodom[i]]))
+  end
+  [(col_names = first.(cols[i]), col_types = last.(cols[i])) for i in 1:length(Ob)]
+end
+
+abstract type AbstractSInstance{CD,AD,Ts} end
+
+struct SInstance{CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple, Idxed,
+                 TablesType <: NamedTuple, IndexType <: NamedTuple} <: AbstractSInstance{CD,AD,Ts}
+  tables::TablesType
+  index::IndexType
+  function SInstance{CD}() where {CD <: CatDesc}
+    SInstance{CD,typeof(AttrDesc(CD())),Tuple{}}()
+  end
+  function SInstance{CD,AD,Ts,Idxed}() where {CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple, Idxed}
+    td = table_desc(CD,AD,Tuple(Ts.parameters))
+    empty_tables = NamedTuple{CD.ob}(
+      [StructArray{NamedTuple{Tuple(td[i].col_names), Tuple{td[i].col_types...}}}(undef,0)
+       for i in 1:length(CD.ob)])
+    index = make_index(CD,AD,Ts,Idxed)
+    new{CD,AD,Ts,Idxed,typeof(empty_tables),typeof(index)}(empty_tables,index)
+  end
+  function SInstance{CD,AD,Ts,Idxed,TT,IT}() where
+    {CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple,Idxed,TT <: NamedTuple,IT <: NamedTuple}
+    SInstance{CD,AD,Ts,Idxed}()
+  end
+  function SInstance{CD,AD,Ts,Idxed,TT,IT}(tables::TT,index::IT) where
+    {CD <: CatDesc, AD <: AttrDesc{CD}, Ts <: Tuple,Idxed,TT <: NamedTuple,IT <: NamedTuple}
+    new{CD,AD,Ts,Idxed,TT,IT}(tables,index)
+  end
+end
+
+
+function fields(::Type{T}) where {T <: NamedTuple}
```

Comment:
I think this custom function can be replaced with the builtin `fieldnames`:

```julia
julia> fieldnames(typeof((x=1,y=2)))
(:x, :y)
```

Reply:
Yeah, sounds good to me, also this reminds me of: https://meta.wikimedia.org/wiki/Cunningham%27s_Law

Reply:
Haha, I didn't know that one but it's so true!